# Obesity Risk Prediction for Cardiovascular Disease

## Introduction
This project aims to predict obesity risk in individuals, a key factor related to cardiovascular disease, using various health-related features. The dataset includes demographic and medical information that will be used to develop a predictive model. The data can be downloaded from the [Kaggle competition page](https://www.kaggle.com/competitions/playground-series-s4e2/data).


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split , GridSearchCV
from sklearn.metrics import classification_report , accuracy_score
from sklearn.preprocessing import LabelEncoder , StandardScaler
from sklearn.ensemble import GradientBoostingClassifier , RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier


In [6]:
train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

In [8]:
train.shape

(20758, 18)

In [10]:
train.head()

,id,Gender,Age,Height,Weight,family_history_with_overweight,FAVC,FCVC,NCP,CAEC,SMOKE,CH2O,SCC,FAF,TUE,CALC,MTRANS,NObeyesdad
0,0,Male,24.443011,1.699998,81.669950,yes,yes,2.000000,2.983297,Sometimes,no,2.763573,no,0.000000,0.976473,Sometimes,Public_Transportation,Overweight_Level_II
1,1,Female,18.000000,1.560000,57.000000,yes,yes,2.000000,3.000000,Frequently,no,2.000000,no,1.000000,1.000000,no,Automobile,Normal_Weight
2,2,Female,18.000000,1.711460,50.165754,yes,yes,1.880534,1.411685,Sometimes,no,1.910378,no,0.866045,1.673584,no,Public_Transportation,Insufficient_Weight
3,3,Female,20.952737,1.710730,131.274851,yes,yes,3.000000,3.000000,Sometimes,no,1.674061,no,1.467863,0.780199,Sometimes,Public_Transportation,Obesity_Type_III
4,4,Male,31.641081,1.914186,93.798055,yes,yes,2.679664,1.971472,Sometimes,no,1.979848,no,1.967973,0.931721,Sometimes,Public_Transportation,Overweight_Level_II


In [12]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20758 entries, 0 to 20757
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              20758 non-null  int64  
 1   Gender                          20758 non-null  object 
 2   Age                             20758 non-null  float64
 3   Height                          20758 non-null  float64
 4   Weight                          20758 non-null  float64
 5   family_history_with_overweight  20758 non-null  object 
 6   FAVC                            20758 non-null  object 
 7   FCVC                            20758 non-null  float64
 8   NCP                             20758 non-null  float64
 9   CAEC                            20758 non-null  object 
 10  SMOKE                           20758 non-null  object 
 11  CH2O                            20758 non-null  float64
 12  SCC                             

In [14]:
train.isnull().sum()

id                                0
Gender                            0
Age                               0
Height                            0
Weight                            0
family_history_with_overweight    0
FAVC                              0
FCVC                              0
NCP                               0
CAEC                              0
SMOKE                             0
CH2O                              0
SCC                               0
FAF                               0
TUE                               0
CALC                              0
MTRANS                            0
NObeyesdad                        0
dtype: int64

In [16]:
test.isnull().sum()

id                                0
Gender                            0
Age                               0
Height                            0
Weight                            0
family_history_with_overweight    0
FAVC                              0
FCVC                              0
NCP                               0
CAEC                              0
SMOKE                             0
CH2O                              0
SCC                               0
FAF                               0
TUE                               0
CALC                              0
MTRANS                            0
dtype: int64

In [18]:
le = LabelEncoder()

In [20]:
obj = train.select_dtypes(include='object')
non_obj = train.select_dtypes(exclude='object')

In [22]:
for i in range(0, obj.shape[1]):
    obj.iloc[:, i] = le.fit_transform(obj.iloc[:, i])

In [24]:
obj = obj.astype("int")

In [26]:
train = pd.concat([obj, non_obj], axis=1)

In [28]:
obj = test.select_dtypes(include='object')
non_obj = test.select_dtypes(exclude='object')
for i in range(0, obj.shape[1]):
    obj.iloc[:, i] = le.fit_transform(obj.iloc[:, i])
obj = obj.astype("int")
test = pd.concat([obj, non_obj], axis=1)

In [30]:
scaler = StandardScaler()

In [32]:
scal_1 = train[['Age']]
scal_2 = train[['Weight']]
train['Age'] = scaler.fit_transform(scal_1)
train['Weight'] = scaler.fit_transform(scal_2)

In [34]:
scal_1 = test[['Age']]
scal_2 = test[['Weight']]
test['Age'] = scaler.fit_transform(scal_1)
test['Weight'] = scaler.fit_transform(scal_2)

In [36]:
x = train.drop(['NObeyesdad','id'],axis=1)
y = train['NObeyesdad']

In [38]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [40]:
lr= LogisticRegression()
gb= GradientBoostingClassifier()
svc= SVC()
gas= GaussianNB()
knn= KNeighborsClassifier()
xb= XGBClassifier()

In [42]:
Algorthim = ['LogisticRegression()',' GradientBoostingClassifier()', 'SVC()', 'GaussianNB()', 'KNeighborsClassifier()', 'XGBClassifier()']
Result_1=[]

In [44]:
def Model(Model):
    Model.fit(x_train,y_train)
    pre = Model.predict(x_test)
    accuracy = accuracy_score(pre, y_test)
    print("Model is: ",Model)
    print("Accuracy is: ",round((accuracy*100),2),"%")
    print(classification_report(pre, y_test))
    Result_1.append(accuracy)

In [46]:
Model(lr)

Model is:  LogisticRegression()
Accuracy is:  80.42 %
              precision    recall  f1-score   support

           0       0.86      0.78      0.82       582
           1       0.67      0.70      0.69       602
           2       0.80      0.77      0.78       564
           3       0.97      0.95      0.96       668
           4       1.00      0.99      1.00       807
           5       0.59      0.64      0.61       447
           6       0.60      0.64      0.62       482

    accuracy                           0.80      4152
   macro avg       0.78      0.78      0.78      4152
weighted avg       0.81      0.80      0.81      4152



C:\Users\zeyil\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [48]:
Model(gb)
Model(svc)
Model(gas)
Model(knn)
Model(xb)

Model is:  GradientBoostingClassifier()
Accuracy is:  90.27 %
              precision    recall  f1-score   support

           0       0.94      0.95      0.94       520
           1       0.88      0.88      0.88       633
           2       0.87      0.88      0.88       537
           3       0.97      0.97      0.97       655
           4       1.00      1.00      1.00       805
           5       0.78      0.78      0.78       486
           6       0.80      0.80      0.80       516

    accuracy                           0.90      4152
   macro avg       0.89      0.89      0.89      4152
weighted avg       0.90      0.90      0.90      4152

Model is:  SVC()
Accuracy is:  85.24 %
              precision    recall  f1-score   support

           0       0.94      0.87      0.90       564
           1       0.77      0.81      0.79       593
           2       0.84      0.82      0.83       557
           3       0.97      0.97      0.97       659
           4       1.00      1.

In [51]:
Final = pd.DataFrame({"Algorithm": Algorthim,
                     "Result: ": Result_1})
print(Final)
print("----------------------------")
print("The Best Algo is: ",Final.max())

                       Algorithm  Result: 
0           LogisticRegression()  0.804191
1   GradientBoostingClassifier()  0.902697
2                          SVC()  0.852360
3                   GaussianNB()  0.663054
4         KNeighborsClassifier()  0.770472
5                XGBClassifier()  0.900289
----------------------------
The Best Algo is:  Algorithm    XGBClassifier()
Result:             0.902697
dtype: object


In [53]:
test_x = test.drop('id', axis=1)

In [55]:
pred_x = xb.predict(test_x)

In [57]:
submission = pd.DataFrame({"id":test['id'],
                           "NObeyesdad":pred_x})

In [59]:
mapping_dict = {
    4: 'Obesity_Type_III',
    3: 'Obesity_Type_II',
    1: 'Normal_Weight',
    2: 'Obesity_Type_I',
    0: 'Insufficient_Weight',
    6: 'Overweight_Level_II',
    5: 'Overweight_Level_I'
}
submission['NObeyesdad'] = submission['NObeyesdad'].map(mapping_dict)

In [61]:
submission.to_csv("submission.csv",index=False)

## Conclusion

In this project, multiple machine learning models were evaluated to predict obesity risk, which is closely related to cardiovascular disease. Various models were tested, including Logistic Regression, Gradient Boosting Classifier, Support Vector Classifier (SVC), Gaussian Naive Bayes, K-Nearest Neighbors (KNN), and XGBoost Classifier. 

- The **XGBoost Classifier** delivered the best performance, achieving an accuracy of **90.03%** on the validation set, and was selected as the final model for the competition. This tree-based model performed exceptionally well in capturing the underlying patterns of the dataset.
- The **Gradient Boosting Classifier** also performed strongly with an accuracy of **90.27%**, indicating that boosting algorithms are highly effective for this task.
- Other models like **Logistic Regression** and **SVC** provided good results with accuracies of **80.42%** and **85.24%**, respectively, but did not perform as well as the boosting models.
- The **Gaussian Naive Bayes** model struggled with an accuracy of **66.31%**, and the **KNN** model, while better, had an accuracy of **77.05%**.

The final Kaggle result of **0.88908** suggests that the model generalizes well to unseen data, though there is still room for improvement. To further refine the model, hyperparameter tuning, feature engineering, and additional techniques such as ensemble methods could be explored. Overall, the XGBoost model demonstrated strong predictive power and is the best model for this task based on both the validation and competition results.
